pkg/shares/parse_compact_shares.go

```diff
@@ -30,79 +26,44 @@ func parseCompactShares(rawShares [][]byte, supportedShareVersions []uint8) (dat
 		}
 	}
 
-	ss := newShareStack(rawShares)
-	return ss.resolve()
-}
-
-// shareStack holds variables for peel
-type shareStack struct {
-	shares  [][]byte
-	dataLen uint64
-	// data may be transactions or intermediate state roots depending
-	// on the namespace ID for this share
-	data   [][]byte
-	cursor int
+	return peel(shares)
 }
 
-func newShareStack(shares [][]byte) *shareStack {
-	return &shareStack{shares: shares}
-}
+// peel parses each unit of data (either a transaction or
+// intermediate state root) and adds it to the underlying slice of data.
+// data may be transactions or intermediate state roots depending
+// on the namespace ID for this share
+func peel(shares []Share) (data [][]byte, err error) {
```

Comment:
@rootulp @evan-forbes please review this function carefully to see if there are some edge cases that we need to consider in it.

Reply:
[optional] I found it difficult to comprehend the recursive `peel` functionality so I wrote a non-recursive implementation. It hasn't been unit tested but was inspired by the implementation in this PR

```
package shares

import (
	"bytes"
	"fmt"
)

// parseCompactShares returns data (transactions or intermediate state roots
// based on the contents of rawShares and supportedShareVersions. If rawShares
// contains a share with a version that isn't present in supportedShareVersions,
// an error is returned. The returned data [][]byte does not have namespaces,
// info bytes, data length delimiter, or unit length delimiters and are ready to
// be unmarshalled.
func parseCompactShares(shares []Share, supportedShareVersions []uint8) (data [][]byte, err error) {
	if len(shares) == 0 {
		return nil, nil
	}
	err = validateShareVersions(shares, supportedShareVersions)
	if err != nil {
		return nil, err
	}

	rawData, err := extractRawData(shares)
	if err != nil {
		return nil, err
	}

	sequenceLen, err := shares[0].SequenceLen()
	if err != nil {
		return nil, err
	}

	if int(sequenceLen) != len(rawData) {
		return nil, fmt.Errorf("sequence length %v does not match the sum of the lengths of the raw data in the shares %v", sequenceLen, len(rawData))
	}

	data, err = parseRawData(rawData)
	if err != nil {
		return nil, err
	}
	return data, nil
}

// validateShareVersions returns an error if the shares contain a share with an
// unsupported share version. Returns nil if all shares contain supported share
// versions.
func validateShareVersions(shares []Share, supportedShareVersions []uint8) error {
	for _, share := range shares {
		infoByte, err := share.InfoByte()
		if err != nil {
			return err
		}
		if !bytes.Contains(supportedShareVersions, []byte{infoByte.Version()}) {
			return fmt.Errorf("unsupported share version %v is not present in the list of supported share versions %v", infoByte.Version(), supportedShareVersions)
		}
	}
	return nil
}

// parseRawData returns the units (transactions, PFB transactions, intermediate
// state roots) contained in raw data by parsing the unit length delimiter
// prefixed to each unit.
func parseRawData(rawData []byte) (units [][]byte, err error) {
	units = make([][]byte, 0)
	for {
		actualData, unitLen, err := ParseDelimiter(rawData)
		if err != nil {
			return nil, err
		}
		if unitLen == 0 {
			return units, nil
		}
		rawData = actualData[unitLen:]
		units = append(units, actualData[:unitLen])
	}
}

// extractRawData returns the raw data contain in the shares. The raw data does
// not contain the namespace ID, info byte, sequence length, or reserved bytes.
func extractRawData(shares []Share) (rawData []byte, err error) {
	for _, share := range shares {
		raw, err := share.RawData()
		if err != nil {
			return nil, err
		}
		rawData = append(rawData, raw...)
	}
	return rawData, nil
}
```

Reply:
> Do you want me to replace my peel function with this 

Sorry for the lack of clarity. I defer to you on which one is clearer but I agree unit tests would increase our confidence on merging either implementation so will let you add tests if you're comfortable doing so.